syntax = "proto3";

package envoy.extensions.dynamic_modules.v3;

import "xds/annotations/v3/status.proto";

import "udpa/annotations/status.proto";
import "validate/validate.proto";

option java_package = "io.envoyproxy.envoy.extensions.dynamic_modules.v3";
option java_outer_classname = "DynamicModulesProto";
option java_multiple_files = true;
option go_package = "github.com/envoyproxy/go-control-plane/envoy/extensions/dynamic_modules/v3;dynamic_modulesv3";
option (udpa.annotations.file_status).package_version_status = ACTIVE;
option (xds.annotations.v3.file_status).work_in_progress = true;

// [#protodoc-title: Dynamic Modules common configuration]

// Configuration of a dynamic module. A dynamic module is a shared object file that can be loaded at various Envoy
// extension points. Currently, only HTTP filter (envoy.filters.http.dynamic_modules) is supported.
//
// All available dynamic modules are loaded at startup. They are not reloaded during the lifetime of the process.
// Once a dynamic module is loaded, it is not unloaded until the process exits.
//
// In Envoy, it searches for dynamic modules in the directory specified by the environment variable
// ``ENVOY_DYNAMIC_MODULES_SEARCH_PATH``. All shared lbraries in the directory are loaded, and they can be
// referenced by a name, following the naming convention of ``lib${name}.so`` as in the standard linker behavior.
// For example, if the dynamic module's file name is ``libfoo.so``, the name is ``foo``.
//
// A module must be compatible with the ABI specified in :repo:`abi.h <source/extensions/dynamic_modules/abi.h>`.
// Currently, compatibility is only guaranteed by an exact version match between the Envoy
// codebase and the dynamic module SDKs. In the future, after the ABI is stabilized, we will revisit
// this restriction and hopefully provide a wider compatibility guarantee. Until then, Envoy
// checks the hash of the ABI header files to ensure that the dynamic modules are built against the
// same version of the ABI.
//
// Currently, the implementation is work in progress and not usable.
message DynamicModuleConfig {
  // The name of the dynamic module. The client is expected to have some configuration indicating where to search for the module.
  // In Envoy, the search path can only be configured via the environment variable ``ENVOY_DYNAMIC_MODULES_SEARCH_PATH``.
  // The actual search path is ``${ENVOY_DYNAMIC_MODULES_SEARCH_PATH}/lib${name}.so``. TODO: make the search path configurable via
  // command line options.
  string name = 1 [(validate.rules).string = {min_len: 1}];
}
